packages/chain-bridge/src/router/BridgeRouter.sol

```diff
@@ -0,0 +1,828 @@
+        ) {
+            operationStatuses[operationId] = status;
+            emit OperationStatusUpdated(operationId, status);
+        } else if (
+            status == BridgeTypes.OperationStatus.FAILED &&
+            operationStatuses[operationId] == BridgeTypes.OperationStatus.SENT
+        ) {
+            operationStatuses[operationId] = status;
+            emit OperationStatusUpdated(operationId, status);
+        }
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function updateReceiveStatus(
+        bytes32 requestId,
+        address recipient,
+        BridgeTypes.OperationStatus status
+    ) external onlyRegisteredAdapter {
+        requestReceivedByAdapter[requestId] = msg.sender;
+
+        // Only update status if it's a failure
+        if (status == BridgeTypes.OperationStatus.FAILED) {
+            operationStatuses[requestId] = status;
+            emit OperationStatusUpdated(requestId, status);
+        }
+
+        // Always emit delivery event
+        emit MessageDelivered(
+            requestId,
+            recipient,
+            status != BridgeTypes.OperationStatus.FAILED
+        );
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function notifyMessageReceived(
+        bytes32 operationId,
+        address asset,
+        uint256 amount,
+        address recipient,
+        uint16 sourceChainId
+    ) external onlyRegisteredAdapter {
+        // Store which adapter received this request
+        requestReceivedByAdapter[operationId] = msg.sender;
+
+        // Emit events for tracking
+        emit MessageDelivered(operationId, recipient, true);
+
+        // If this is a transfer, emit the transfer event
+        if (asset != address(0) && amount > 0) {
+            emit TransferReceived(
+                operationId,
+                asset,
+                amount,
+                recipient,
+                sourceChainId
+            );
+        }
+    }
+
+    /// @inheritdoc IBridgeRouter
+    function deliverReadResponse(
+        bytes32 operationId,
+        uint16 sourceChainId,
+        bytes calldata resultData
+    ) external nonReentrant onlyRegisteredAdapter {
+        if (operationToAdapter[operationId] != msg.sender) {
+            revert Unauthorized();
+        }
+
+        address originator = readRequestToOriginator[operationId];
+        if (originator == address(0)) revert InvalidParams();
+
+        // Try to deliver the response
+        bool delivered = false;
+
+        // Check if the originator implements the ICrossChainStateReadReceiver interface
+        bytes4 interfaceId = type(ICrossChainStateReadReceiver).interfaceId;
+        try
+            ICrossChainStateReadReceiver(originator).supportsInterface(
+                interfaceId
+            )
+        returns (bool supported) {
+            if (supported) {
+                try
+                    ICrossChainStateReadReceiver(originator).receiveStateRead(
```

Comment:
there's an inconsistency on how the callbacks are called.

`receiveStateRead` is called from within the `BridgeRouter` after a call from adapter `deliverReadResponse(..)`
`receiveMessage` is called directly from the adapter, same as 
`receiveMessageWithAssets` is being called directly from the adapter